File: CFD_Principles_and_Applications_Blazek/BookCode/C++/unstructured_2d/src/irsmoo.cpp
/// @file irsmoo.cpp
///
/// Central implicit residual smoothing.
///
//*****************************************************************************
//
//  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
//  Created February 15, 2014
//  Last modification: July 12, 2014
// 
//=============================================================================
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//*****************************************************************************

#include "timeDiscr.h"

/// Applies the central implicit smoothing method to the residuals
/// by using Jacobi iteration.
///
/// @param geometry    geometrical data
/// @param spaceDiscr  spatial discretization
///
void TimeDiscr::Irsmoo( const Geometry &geometry, SpaceDiscr &spaceDiscr )
{
  int  itirs, i, j, ie;
  REAL den;

  CONSVARS       *rhs  = spaceDiscr.rhs;
  Geometry::EDGE *edge = geometry.edge;

  // initialize counter (no. of contributions to a node);
  // store the non-smoothed residual in "rhsOld"

  for (i=0; i<geometry.nNodes; i++)
  {
    nContr[i] = 0;
    rhsOld[i] = rhs[i];
  }

  // Jacobi iteration *********************************************************

  for (itirs=1; itirs<=nIterIrs; itirs++)
  {
    // zero out nodal contributions

    for (i=0; i<geometry.nndInt; i++)
    {
      rhsIter[i].dens = 0.0;
      rhsIter[i].xmom = 0.0;
      rhsIter[i].ymom = 0.0;
      rhsIter[i].ener = 0.0;
    }

    // loop over edges: first iteration => set counter

    if (itirs == 1)
    {
      for (ie=0; ie<geometry.nedInt; ie++)
      {
        i = edge[ie].i;
        j = edge[ie].j;
        nContr[i]++;
        nContr[j]++;

        rhsIter[i].dens += rhs[j].dens;
        rhsIter[i].xmom += rhs[j].xmom;
        rhsIter[i].ymom += rhs[j].ymom;
        rhsIter[i].ener += rhs[j].ener;

        rhsIter[j].dens += rhs[i].dens;
        rhsIter[j].xmom += rhs[i].xmom;
        rhsIter[j].ymom += rhs[i].ymom;
        rhsIter[j].ener += rhs[i].ener;
      }

      // sum up no. of contributions at periodic nodes
      BndConds::Periodic( geometry,nContr );
    }

    // loop over edges: without the counter

    else
    {
      for (ie=0; ie<geometry.nedInt; ie++)
      {
        i = edge[ie].i;
        j = edge[ie].j;

        rhsIter[i].dens += rhs[j].dens;
        rhsIter[i].xmom += rhs[j].xmom;
        rhsIter[i].ymom += rhs[j].ymom;
        rhsIter[i].ener += rhs[j].ener;

        rhsIter[j].dens += rhs[i].dens;
        rhsIter[j].xmom += rhs[i].xmom;
        rhsIter[j].ymom += rhs[i].ymom;
        rhsIter[j].ener += rhs[i].ener;
      }
    }

    // periodic boundaries

    BndConds::Periodic( geometry,rhsIter );

    // new smoothed residual

    for (i=0; i<geometry.nndInt; i++)
    {
      den         = 1.0/(1.0+((REAL)nContr[i])*epsIrs);
      rhs[i].dens = (rhsIter[i].dens*epsIrs+rhsOld[i].dens)*den;
      rhs[i].xmom = (rhsIter[i].xmom*epsIrs+rhsOld[i].xmom)*den;
      rhs[i].ymom = (rhsIter[i].ymom*epsIrs+rhsOld[i].ymom)*den;
      rhs[i].ener = (rhsIter[i].ener*epsIrs+rhsOld[i].ener)*den;
    }
  } // itirs
}

File: CFD_Principles_and_Applications_Blazek/BookCode/Fortran/unstructured_2d/src/irsmoo.f90
!> @file irsmoo.f90
!!
!! Central implicit residual smoothing.
!
! *****************************************************************************
!
!  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
!  Created February 25, 2014
!  Last modification: June 3, 2014
!
! *****************************************************************************
!
!  This program is free software; you can redistribute it and/or
!  modify it under the terms of the GNU General Public License
!  as published by the Free Software Foundation; either version 2
!  of the License, or (at your option) any later version.
!
!  This program is distributed in the hope that it will be useful,
!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
!  GNU General Public License for more details.
!
!  You should have received a copy of the GNU General Public License
!  along with this program; if not, write to the Free Software
!  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
!
! *****************************************************************************

!> Applies the central implicit smoothing method to the residuals
!! by using Jacobi iteration.
!!
!! @param ncontr  number of contributions to a node
!! @param rhsold  storage for the initial non-smoothed residual
!! @param rhsit   nodal contributions to the residual during the iteration
!!
subroutine Irsmoo( ncontr,rhsold,rhsit )

  use ModDataTypes
  use ModGeometry
  use ModNumerics
  use ModPhysics
  use ModInterfaces, only : Periodic
  implicit none

! parameters
  integer     :: ncontr(:)
  real(rtype) :: rhsold(:,:), rhsit(:,:)

! local variables
  integer     :: itirs, i, j, ie, ib, ibn, ibegn, iendn
  real(rtype) :: den

! *****************************************************************************
! initialize counter (no. of contributions to a node);
! store the non-smoothed residual in "rhsold"

  do i=1,nndint
    ncontr(i)   = 0
    rhsold(1,i) = rhs(1,i)
    rhsold(2,i) = rhs(2,i)
    rhsold(3,i) = rhs(3,i)
    rhsold(4,i) = rhs(4,i)
  enddo

! Jacobi iteration ------------------------------------------------------------

  do itirs=1,nitirs

! - zero out nodal contributions

    do i=1,nndint
      rhsit(1,i) = 0.D0
      rhsit(2,i) = 0.D0
      rhsit(3,i) = 0.D0
      rhsit(4,i) = 0.D0
    enddo

! - loop over edges - first iteration => set counter

    if (itirs == 1) then
      do ie=1,nedint
        i          = edge(1,ie)
        j          = edge(2,ie)
        ncontr(i)  = ncontr(i) + 1
        ncontr(j)  = ncontr(j) + 1
        rhsit(1,i) = rhsit(1,i) + rhs(1,j)
        rhsit(2,i) = rhsit(2,i) + rhs(2,j)
        rhsit(3,i) = rhsit(3,i) + rhs(3,j)
        rhsit(4,i) = rhsit(4,i) + rhs(4,j)

        rhsit(1,j) = rhsit(1,j) + rhs(1,i)
        rhsit(2,j) = rhsit(2,j) + rhs(2,i)
        rhsit(3,j) = rhsit(3,j) + rhs(3,i)
        rhsit(4,j) = rhsit(4,j) + rhs(4,i)
      enddo

! - sum up no. of contributions at periodic nodes

      ibegn = 1
      do ib=1,nsegs
        iendn = ibound(2,ib)
        if (btype(ib)>=700 .and. btype(ib)<800) then
          do ibn=ibegn,iendn
            i         = bnode(1,ibn)
            j         = bnode(2,ibn)
            ncontr(i) = ncontr(i) + ncontr(j)
            ncontr(j) = ncontr(i)
          enddo
        endif
        ibegn = iendn + 1
      enddo

! - loop over edges - without counter

    else
      do ie=1,nedint
        i          = edge(1,ie)
        j          = edge(2,ie)
        rhsit(1,i) = rhsit(1,i) + rhs(1,j)
        rhsit(2,i) = rhsit(2,i) + rhs(2,j)
        rhsit(3,i) = rhsit(3,i) + rhs(3,j)
        rhsit(4,i) = rhsit(4,i) + rhs(4,j)

        rhsit(1,j) = rhsit(1,j) + rhs(1,i)
        rhsit(2,j) = rhsit(2,j) + rhs(2,i)
        rhsit(3,j) = rhsit(3,j) + rhs(3,i)
        rhsit(4,j) = rhsit(4,j) + rhs(4,i)
      enddo
    endif    ! itirs > 1

! - periodic boundaries

    call Periodic( rhsit )

! - new smoothed residual

    do i=1,nndint
      den      = 1.D0/(1.D0+epsirs*Real(ncontr(i)))
      rhs(1,i) = (rhsit(1,i)*epsirs+rhsold(1,i))*den
      rhs(2,i) = (rhsit(2,i)*epsirs+rhsold(2,i))*den
      rhs(3,i) = (rhsit(3,i)*epsirs+rhsold(3,i))*den
      rhs(4,i) = (rhsit(4,i)*epsirs+rhsold(4,i))*den
    enddo

  enddo    ! loop over itirs

end subroutine Irsmoo
